subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/cell.go

```diff
@@ -211,58 +248,78 @@ func (c *PhysicalCell) SetReserved(reserved bool) {
 	c.reserved = reserved
 }
 
+func (c *PhysicalCell) GetStatus() *api.PhysicalCellStatus {
+	return c.status
+}
+
 // VirtualCell defines a cell in a VC.
 type VirtualCell struct {
 	GenericCell
 	vc                   api.VirtualClusterName // name of its VC
 	rid                  api.ReservationId      // reservation ID
-	indexInChain         int32                  // index of the cell in the ChainCellList it belongs to (assigned in initialization)
 	preAssignedCell      *VirtualCell           // top level cell of this cell chain
 	physicalCell         *PhysicalCell          // points to the bound physical cell
 	preBoundPhysicalCell *PhysicalCell          // points to the temporarily bound physical cell (before the binding is confirmed)
+	status               *api.VirtualCellStatus // json representation of the status of this cell to expose to users
 }
 
-func NewVirtualCell(vc api.VirtualClusterName,
+func NewVirtualCell(
+	vc api.VirtualClusterName,
 	c CellChain,
 	l CellLevel,
 	g bool,
 	n int32,
-	i int32,
-	pac *VirtualCell) *VirtualCell {
+	pac *VirtualCell,
+	cellType string,
+	address string) *VirtualCell {
 
 	return &VirtualCell{
 		GenericCell: GenericCell{
 			chain:                  c,
 			level:                  l,
 			priority:               freePriority,
+			address:                address,
 			atOrHigherThanNode:     g,
 			totalGpuNum:            n,
 			usedGpuNumAtPriorities: map[CellPriority]int32{},
 		},
 		vc:              vc,
-		indexInChain:    i,
 		preAssignedCell: pac,
+		status: &api.VirtualCellStatus{
+			CellStatus: api.CellStatus{
+				CellType:    cellType,
+				CellAddress: address,
+				State:       api.FreeState,
+				Priority:    int32(freePriority),
+			},
+		},
 	}
 }
 
-func (c *VirtualCell) GetName() string {
-	str := string(c.chain)
-	if c.rid != "" {
-		str = string(c.rid)
+func (c *VirtualCell) SetPriority(p CellPriority) {
+	c.priority = p
+	c.status.Priority = int32(p)
+	state := api.UsedState
+	if p == freePriority {
+		state = api.FreeState
+	}
+	c.status.State = state
+	if c.status.PhysicalCell != nil {
```

Comment:
If the old state is bad, will it override it to be others?

Reply:
Solved as we have used separated states for bad and used